aten/src/ATen/native/cpu/SpmmReduceKernel.cpp

```diff
@@ -0,0 +1,511 @@
+
+  TORCH_CHECK(reduce == ReductionType::MAX || reduce == ReductionType::MIN);
+  int64_t nnz = values.numel();
+  if (nnz == 0) {
+    return;
+  }
+
+  auto other = other_.contiguous();
+
+  scalar_t* out_data = out.data_ptr<scalar_t>();
+  index_t* arg_out_data = arg_out.data_ptr<index_t>();
+  auto csr_data = crow_indices.accessor<index_t, 1>();
+  auto col_data = col_indices.accessor<index_t, 1>();
+  auto val_data = values.accessor<scalar_t, 1>();
+  scalar_t* other_data = other.data_ptr<scalar_t>();
+
+  int64_t M = crow_indices.numel() - 1;
+  int64_t K = other.size(-1);
+
+  at::parallel_for(0, M, 1, [&](int64_t begin, int64_t end) {
+    int64_t row_start, row_end, c;
+    for (const auto m : c10::irange(begin, end)) {
+      row_start = csr_data[m];
+      row_end = csr_data[m + 1];
+
+      scalar_t* out_ptr = out_data + m * K;
+      index_t* arg_out_ptr = arg_out_data + m * K;
+
+      int64_t count = row_end - row_start;
+      if (count != 0) {
+        init<scalar_t, reduce>(out_ptr, K, /*include_self*/false);
+        for (const auto e : c10::irange(row_start, row_end)) {
+          c = col_data[e];
+          scalar_t val = val_data[e];
+
+          scalar_t* other_ptr = other_data + c * K;
+          for (const auto k : c10::irange(K)) {
+            update_with_index<scalar_t, index_t, reduce>(
+                &out_ptr[k], val *  other_ptr[k], &arg_out_ptr[k], index_t(e));
+          };
+        }
+      }
+    }
+  });
+}
+
+template <typename scalar_t, typename index_t, ReductionType reduce>
+void spmm_reduce_backward_input_kernel_impl(
+    const Tensor& grad_self,
+    const Tensor& grad_out_,
+    const Tensor& crow_indices,
+    const Tensor& col_indices,
+    const Tensor& other_,
+    const Tensor& row_indices) {
+
+  int64_t nnz = grad_self._nnz();
+  if (nnz == 0) {
+    return;
+  }
+
+  auto grad_out = grad_out_.contiguous();
+  auto other = other_.contiguous();
+
+  auto values = grad_self.values();
+  auto grad_values_data = values.accessor<scalar_t, 1>();
+  scalar_t* grad_out_data = grad_out.data_ptr<scalar_t>();
+  auto crow_data = crow_indices.accessor<index_t, 1>();
+  auto col_data = col_indices.accessor<index_t, 1>();
+  scalar_t* other_data = other.data_ptr<scalar_t>();
+  auto row_data = row_indices.accessor<index_t, 1>();
+
+  int64_t K = grad_out.size(1);
+
+  using Vec = vec::Vectorized<vec::vec_scalar_t<scalar_t>>;
+  at::parallel_for(0, nnz, 1, [&](int64_t begin, int64_t end) {
+    for (const auto i : c10::irange(begin, end)) {
+      index_t row = row_data[i], col = col_data[i];
+
+      scalar_t val = vec::map2_reduce_all<scalar_t>(
+          [](Vec x, Vec y) { return x * y; },
+          [](Vec x, Vec y) { return x + y; },
+          other_data + col * K,
+          grad_out_data + row * K,
+          K);
+
+      if (reduce == ReductionType::MEAN) {
+        index_t row_start = crow_data[row], row_end = crow_data[row + 1];
+        val /= std::max((index_t)1, row_end - row_start);
```

Comment:
No... this just reminds me that the `csc` support in this PR is problematic: we can't simply do a transpose upon `csc` and do the following computation, logically it would be different from `torch.mm()` where

No matter what the layout is (`strided`, `sparse_coo`, `sparse_csr`, `sparse_csc`), `torch.mm()` should have the identical result. But here we are computing:
* `A^T * B` if A is csc
* `A * B` if A is csr

So i think we should disable CSC support for now, (maybe add more layout support in the future).